src/rulelib.cpp

```diff
@@ -917,74 +870,131 @@ ruleset_print(ruleset_t *rs, rule_t *rules, int detail)
-	if (detail)
-		rule_vector_print(r->truthtable, n);
+	if (detail) {
+        printf(":");
+		rule_vector_print(r->truthtable, nsamples);
+    }
+    else
+	    printf("\n");
 }
 
 void
-rule_vector_print(VECTOR v, int n)
+rule_vector_print(VECTOR v, int nsamples)
 {
 #ifdef GMP
-	mpz_out_str(stdout, 16, v);
-	printf("\n");
+	char* str = mpz_get_str(NULL, 2, v);
+    int len = strlen(str);
+    for(int i = 0; i < (nsamples - len); i++) {
+        fputc('0', stdout);
+    }
+    fputs(str, stdout);
+	fputc('\n', stdout);
 #else
-	for (int i = 0; i < n; i++)
+	/*for (int i = 0; i < n; i++)
 		printf("0x%lx ", v[i]);
-	printf("\n");
+	printf("\n");*/
+
+    v_entry m = ~(((v_entry) -1) >> 1);
+    unsigned n = (nsamples + BITS_PER_ENTRY - 1) / BITS_PER_ENTRY;
+    for(unsigned i = 0; i < n; i++) {
+        v_entry t = v[i];
+        for(unsigned j = 0; j < BITS_PER_ENTRY && (i * BITS_PER_ENTRY + j) < (unsigned)nsamples; j++) {
+            printf("%d", !!(t & m));
+            t <<= 1;
+        }
+    }
+    fputc('\n', stdout);
 #endif
 
 }
 
 void
-rule_print_all(rule_t *rules, int nrules, int nsamples)
+rule_print_all(rule_t *rules, int nrules, int nsamples, int detail)
 {
-	int i, n;
+	int i;
 
-	n = (nsamples + BITS_PER_ENTRY - 1) / BITS_PER_ENTRY;
 	for (i = 0; i < nrules; i++)
-		rule_print(rules, i, n, 1);
+		rule_print(rules, i, nsamples, detail);
 }
 
 /*
  * Return 0 if bit e is not set in vector v; return non-0 otherwise.
  */
 int
-rule_isset(VECTOR v, int e) {
+rule_isset(VECTOR v, int e, int n) {
 #ifdef GMP
 	return mpz_tstbit(v, e);
 #else
-	return ((v[e/BITS_PER_ENTRY] & (1 << (e % BITS_PER_ENTRY))) != 0);
+    e = -e + n - 1;
+    if(e >= n) {
+        return 0;
+    }
+
+    v_entry one = 1;
+    int shift = BITS_PER_ENTRY - (e % BITS_PER_ENTRY) - 1;
+/*    if(e / BITS_PER_ENTRY == ((n + BITS_PER_ENTRY - 1) / BITS_PER_ENTRY - 1) && (n % BITS_PER_ENTRY) != 0) {
+        shift -= BITS_PER_ENTRY - (n % BITS_PER_ENTRY);
+    }
+*/
+	return !!(v[e / BITS_PER_ENTRY] & (one << shift));
+#endif
+}
+
+void
+rule_set(VECTOR v, int e, int val, int n) {
+#ifdef GMP
+    if(val)
+	    mpz_setbit(v, e);
+    else
+        mpz_clrbit(v, e);
+#else
+    e = -e + n - 1;
+    if(e >= n) {
+        return;
+    }
+
+    v_entry one = 1;
+    int shift = BITS_PER_ENTRY - (e % BITS_PER_ENTRY) - 1;
+    // Only needed if the last v_entry is not full and right-aligned
+/*    if(e / BITS_PER_ENTRY == ((n + BITS_PER_ENTRY - 1) / BITS_PER_ENTRY - 1) && (n % BITS_PER_ENTRY) != 0) {
+        shift -= BITS_PER_ENTRY - (n % BITS_PER_ENTRY);
+    }
+*/
+    if(val)
+        v[e / BITS_PER_ENTRY] |= (one << shift);
+    else
+	    v[e/BITS_PER_ENTRY] &= ((v_entry) -1) - (one << shift);
 #endif
 }
 
 int
-rule_vector_equal(const VECTOR v1, const VECTOR v2, short len1, short len2) {
+rule_vector_cmp(const VECTOR v1, const VECTOR v2, int len1, int len2) {
 #ifdef GMP
-    return !mpz_cmp(v1, v2);
+    return mpz_cmp(v1, v2);
```

Comment:
Whoaa this is a big change -- did you make sure to find everywhere in the code we used it?

Reply:
Also, why the change?

Reply:
The only places rule_vector_equal (which I renamed to rule_vector_cmp) was used were lines 61 and 63 of pmap.hh. I made sure to change those. The reason for the change was because I needed a comparison function that returned 0, -1, or 1 depending on which of two vectors was greater (I used it in the minority bound generation), and since rule_vector_equal was essentially what I needed, I renamed it to rule_vector_cmp and made it return 0, -1, or 1.